### ISSUE TITLE
Physical mapped button has double function

### ISSUE THREAD

**Issue by RyanBram** [hrydgard/ppsspp#7254]
Hi, PPSSPP developers,
Recently I bought a Xbox style gamepad and successfully pairing it to my Android device.
I could map every button in PPSSPP, but found minor issue.

In my device, PPSSPP recognize **SELECT** button as **BACK** button from Android. Although I can map this button, but PPSSPP still recognize it as **BACK** button, therefore when I press the button it will execute two functions: As **SELECT** button and for showing PPSSPP pause menu. It isn't comfort to always return to the game everytime I need to press **SELECT** button.

I really hope you want to fix this issue before 1.0 release by ignoring every Android function that already mapped in PPSSPP. In ePSXe this is possible, and I hope PPSSPP also.

*Best regards,
RyanBram*

**Comment by Bigpet** [comment]
I think the issue was that the driver pretty much just send a "Back" command.

Last I heard @hrydgard was not very enthusiastic about disabling the "Back" command, because it's not intuitive and leads to unexpected behavior.

I mean you could theoretically delay the reaction to the back command and then cancel it if a button press is detected but that all seems very hacky to me, can't you just map one of the triggers to the select button?

**Comment by hrydgard** [comment]
That might do the trick, try the latest build when it's done.

**Comment by RyanBram** [comment]
Hi, Henrik,

Thanks for the patch. I tried with two of my Android device. First is my HP 7 Voice Tab and second is in my Chinese Android TV Box, Tronsmart. Unfortunately both device still recognize **SELECT** button as **BACK** button.

I don't know much about programming, but it seems your patch only affect to PPSSPP itself while Android still recognize it as **BACK** button. It is very strange as when I am trying to map the button, the **SELECT** button can be recognized well without going back or canceling the mapping process (as **BACK** button . This problem only appear when I am playing the game.

*Best regards,
RyanBram*

**Comment by hrydgard** [comment]
Hm, alright. I'll look it over again...

**Comment by hrydgard** [comment]
Alright, made a new commit that will display details of the last keyboard event on the Analog Test screen that you can reach from control settings / control mapping.

I need to know the Device ID that you see there when you press the back button, can you check?

**Comment by RyanBram** [comment]
Hi, Henrik,

I tried your new build and this is the result:
Keycode: 4 Device ID: 0
That's result is same when I press my Android native  **BACK** button.

As additional info, my Android seems recognize my controller as XBox 360 controller. I knew that when I am playing with ePSXe and the emulator show what kind of controller conected to my Android.

*Regards,
RyanBram*

**Comment by hrydgard** [comment]
Do other buttons on your pad produce Device ID: 0, or just the back button?

**Comment by hrydgard** [comment]
Alright, one more attempt..

**Comment by RyanBram** [comment]
Ok. I'll try and report the result to you...

**Comment by RyanBram** [comment]
Hi, Henrik,

Sorry for make you waiting so long.
The following is the detailed list of my Gamepad layout:

    =================================
    PPSSPP
    =================================
    UP    = Keycode: 19 Device ID:  1 (pad1.Y HAT-)
    Down  = Keycode: 20 Device ID:  1 (pad1.Y HAT+)
    Left  = Keycode: 21 Device ID:  1 (pad1.X HAT-)
    Right = Keycode: 22 Device ID:  1 (pad1.X HAT+)
    O     = Keycode: 97 Device ID: 10 (pad1.[B])
    X     = Keycode: 96 Device ID: 10 (pad1.[A])
    []    = Keycode: 99 Device ID: 10 (pad1.[X])
    /\    = Keycode:100 Device ID: 10 (pad1.[Y])
    Start = Keycode:108 Device ID: 10 (pad1.Start)
    Select= Keycode:  4 Device ID:  0 (built-in.Back)
    L1    = Keycode:102 Device ID: 10 (pad1.[L1])
    R1    = Keycode:103 Device ID: 10 (pad1.[L2])
    L2    = Undetected                (pad1.TriggerL+)
    R2    = Undetected                (undetected)
    L3    = Keycode:106 Device ID: 10 (pad1.ThumbL)
    R3    = Keycode:107 Device ID: 10 (pad1.ThumbR)
    Analog Up    =                    (pad1.Y Axis-)
    Analog Down  =                    (pad1.Y Axis+)
    Analog Left  =                    (pad1.X Axis-)
    Analog Right =                    (pad1.X Axis+)

Compared to PPSSPP, ePSXe will detect my Gamepad layout as:

    =================================
    ePSXe
    =================================
    UP           = (Axis Hat Y)
    Down         = (Axis Hat Y)
    Left         = (Axis Hat X)
    Right        = (Axis Hat X)
    O            = (Button B)
    X            = (Button A)
    []           = (Button X)
    /\           = (Button Y)
    Start        = (Button Start)
    Select       = (Back)
    L1           = (Button L1)
    R1           = (Button R2)
    L2           = (Axis LTrigger)
    R2           = (Axis RTrigger)
    L3           = (Button ThumbL)
    R3           = (Button ThumbR)
    Analog Up    = (Axis Y)
    Analog Down  = (Axis Y)
    Analog Left  = (Axis X)
    Analog Right = (Axis X)

Some conclusion from me:
* PPSSPP and ePSXe have opposite method when detecting D-Pad and Analog Pad input
* In ePSXe, built-in *BACK* button still has function as *BACK* button, while mapped *SELECT* button has function as *SELECT* button even though ePSXe recognize it as *BACK* button. In PPSSPP, both button (built-in and gamepad) has no difference.

For additional information, my gamepad support X-Input, and ePSXe will detect it as XBox 360 Controller. The appearance of my gamepad is as follow:
![Tronsmart Controller](http://www.tronsmart.com/images/stories/virtuemart/product/1-pic.jpg)

Sorry for any inconvenience.

Best regards,
RyanBram

**Comment by hrydgard** [comment]
Oh you are still getting device-id == 0 for back with the very latest builds? I thought that would have fixed it, darn. Thanks for the detailed report, I'll investigate some more.

**Comment by jackjt8** [comment]
PPSSPP v1.0.1 | Nexus 5 Android 5.1.1 | Moga Pro Power (Both modes)

Not only does SELECT act as device.id == 0, but any button I map to SELECT shows similar behaviours. By default SELECT and Right Stick Button (?) act as SELECT/device.id == 0. I've tried mapping L2/R2 to act as SELECT, but it still ends up working as device.id == 0. 

If you need more info, just let me know.

**Comment by hong620** [comment]
PPSSPP v1.4.2-429-g49c92c69c | Galaxy S3 LTE Android 4.4.4(API LEVEL 19) | XBOX One 3rd Gen Pad(aka S pad) as Android default Bluetooth pairing

seems still got issues in Select key as 'built-in.Back' problem.
i've try again as 'Empty Null in SELECT and PPSSPP Pause' but still it's force to pop-up ppsspp pause menu.

try other emulator or Frontend like ePSXe, Drastic, RetroArch are free from this issue.
it's not just recently issues, i've still remeber this issue was in 1.4.x period.

maybe this issue shown old api level Android versions, but seriously need to investigate and research with XBone pad(as 3rd gen BT on Android)

**Comment by hong620** [comment]
@hrydgard 

try tested oldest possible to download from automatic build bot, 1.4.2-0 one.
still 'built-in.Back' button forced lock in to Pause Menu call button.

one thing strange thing is, acctual set-in 'pause' button is only 'load pause menu' function, not functiong as 'leave pause menu'.
but malfunctioned select key pause menu call is functiong as both 'load' and 'leave' pause menu.

also even in 'android pad's built-in.Back key' malfunction as 'pause menu call',
if you're set normally as 'select button', it's acctually 'functioning as select key' call but same as lighting period of time it's functioning as 'pauce menu call'.

**so it's call pause menu before functioning as select key call** and then if you're keep pressing that button, then it's will infinitly functioning as button call and randomly functiong as select key and same times that infinitly menu load/leave will make app slower and make entire android os almost hang, never be recover.

**Comment by ds22x** [comment]
Seems like this is still an issue, at least on Android TV using X-input controllers.
It would be nice if, seeing as you can actually map a button to pause emulation, to remove the hard coded part and set the "pause" button mapping to "build-in back" as default instead.

**Comment by Zenijin** [comment]
Yeah, this is definitely still an issue. Weirdly, even if I use another app to remap the select button on my controller to do something else, and have it assigned only to the SELECT function in PPSSPP, it STILL opens the system menu (albeit only for a second, and then closes itself).

**Comment by timpauvictor** [comment]
Yup still an issue for me.
I'm surprised more people aren't complaining considering X-input controllers are so commonplace nowadays

**Comment by hrydgard** [comment]
Alright, I've decided to investigate and get this fixed. What XInput controller should I get?

I already have an Xbox 360 controller with the receiver thingy for PC, will that work on Android?

**Comment by babywin** [comment]
> Alright, I've decided to investigate and get this fixed. What XInput controller should I get?
> 
> I already have an Xbox 360 controller with the receiver thingy for PC, will that work on Android?

Thank you very much.
I think a wired Xbox 360 controller will be the best choice. I'm personally using one of those.

**Comment by Zenijin** [comment]
8bitdo controllers are generally a popular Android option. I'm using a Gamesir X2, which is a pretty awesome phone controller.

**Comment by hrydgard** [comment]
Actually I don't think I need to get a new controller, the 360 wireless adapter actually works on the Shield no problem. And I'm having the same issue. 

As far as I can tell it seems to be Android overriding the message into a generic Back instead of pad Back/Select. It also takes over the function of the central Xbox button - that one just exists to Android TV main menu, not sure I can do anything about that.

But need to somehow ensure that even if you map Back to something, there's always a way to get to the menu from in-game. Maybe scan the mappings and make sure some pad button is mapped...

**Comment by hrydgard** [comment]
Hm, I had another idea. Switching away from PPSSPP and back can always be done. It might be useful to always trigger the pause screen when that happens - then that would be an emergency way out. I guess it might potentially annoy some people though who like to switch between apps a lot - but it could be an Android-TV-only feature for example..

**Comment by babywin** [comment]
> Actually I don't think I need to get a new controller, the 360 wireless adapter actually works on the Shield no problem. And I'm having the same issue.
> 
> As far as I can tell it seems to be Android overriding the message into a generic Back instead of pad Back/Select. It also takes over the function of the central Xbox button - that one just exists to Android TV main menu, not sure I can do anything about that.
> 
> But need to somehow ensure that even if you map Back to something, there's always a way to get to the menu from in-game. Maybe scan the mappings and make sure some pad button is mapped...

Yeah! I think any xinput controller will do it. You can continue with your Xbox 360 wireless.
I don't think people do like to switch between apps so often.

**Comment by Zenijin** [comment]
> I don't think people do like to switch between apps so often.
Depends. A lot of the time I'll be talking to people on Discord and I'll keep switching between game and Discord. I imagine that's relatively common.

**Comment by unknownbrackets** [comment]
So maybe...
 * If Back is mapped to something (i.e. Select)...
 * And, nothing is mapped to Pause

Then:
 * Switching from another app to PPSSPP always results in PPSSPP being paused.
 * Switching from/to PPSSPP is the *only* way to pause PPSSPP.

And otherwise, behavior would stay the same as now (i.e. switching would not cause pause and you could just press the button)?

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, I guess that's pretty good. Avoids adding yet another option.

Hopefully Android doesn't helpfully eat Back key-ups, or something, in the process of destroying its identity by forcing the device ID to not be gamepad...

**Comment by hrydgard** [comment]
Another annoyance is that it seems XInput controllers on Android output Up, Left, Down, Right keys instead of HAT X/Y -/+ analog events, as some other controllers do. So on Android we should probably default to supporting both, a common complaint is that the D-pad isn't mapped properly by default.

**Comment by hrydgard** [comment]
Yet another thing, since I'll already be messing with the defaults, I'm thinking about a new good default for Back on the gamepad.

Clicking the left stick is too easy to do by accident. Clicking the right stick maybe? Or even left analog trigger? (not shoulder button, that one is mapped to PSP 'L').

**Comment by unknownbrackets** [comment]
Since Guide is taken, left analog trigger makes sense, I think.  I don't think anyone will guess right stick click.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, left analog trigger is a lot easier to find by button-mashing heh.

**Comment by KolJaY** [comment]
@hrydgard 
Thank you for all the work you have done so far!
Sincd yo are in fixing mappins issues of the app I would like to inform you that this problem presents also in the xbox one adaptation of the app. (E.g. pressing B button causes PPSSPP menu to appear)

**Comment by hrydgard** [comment]
With https://github.com/hrydgard/ppsspp/pull/14718 merged, much of this issue should be fixed - at least Back should be properly mappable to Select.

I am going to check UWP on PC, but I don't have an Xbox to test on.

**Comment by hrydgard** [comment]
It seems to work correctly on UWP on PC. Not gonna do more here for 1.12.
